internal/app/singularity/pull.go

```
@@ -0,0 +1,269 @@
+)
+
+// LibraryPull will download the image specified by file from the library specified by libraryURI.
+// After downloading, the image will be checked for a valid signature and removed if it does not contain one,
+// unless specified not to by the unauthenticated bool
+func LibraryPull(name, ref, transport, fullURI, libraryURI, keyServerURL, authToken string, force, unauthenticated bool) error {
+	if !force {
+		if _, err := os.Stat(name); err == nil {
+			return fmt.Errorf("image file already exists: %q - will not overwrite", name)
+		}
+	}
+
+	libraryImage, err := client.GetImage(libraryURI, authToken, fullURI)
+	if err != nil {
+		return fmt.Errorf("while getting image info: %v", err)
+	}
+
+	// required in order to properly allow for library pulls without transport in uri
+	// otherwise uri becomes malformed see https://github.com/sylabs/singularity/pull/2683
+	var imageName string
+	if transport == "" {
+		imageName = uri.GetName("library://" + fullURI)
+	} else {
+		imageName = uri.GetName(fullURI)
+	}
+	imagePath := cache.LibraryImage(libraryImage.Hash, imageName)
+	exists, err := cache.LibraryImageExists(libraryImage.Hash, imageName)
+	if err != nil {
+		return fmt.Errorf("unable to check if %v exists: %v", imagePath, err)
+	}
+	if !exists {
+		sylog.Infof("Downloading library image")
+		if err = client.DownloadImage(imagePath, fullURI, libraryURI, true, authToken); err != nil {
+			return fmt.Errorf("unable to Download Image: %v", err)
+		}
+
+		if cacheFileHash, err := client.ImageHash(imagePath); err != nil {
+			return fmt.Errorf("error getting imagehash: %v", err)
+		} else if cacheFileHash != libraryImage.Hash {
+			return fmt.Errorf("cached file hash(%s) and expected hash(%s) does not match", cacheFileHash, libraryImage.Hash)
+		}
+	}
+
+	// Perms are 777 *prior* to umask
+	dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+	if err != nil {
+		return fmt.Errorf("%v", err)
```

Can you just return the err? Instead of passing it thrught `fmt.Errorf()`?

```suggestion
		return err
```